### DISCUSSION TITLE
[WASM] WebView2

### DISCUSSION THREAD

**Discussion by DierkDroth** [unoplatform/uno#7739] [General]
- What's the official recommendation for migrating WinUI WebView2 code to WASM? What is it I could do/implement right now?
- Will there be an UNO WASM wrapper around some native browser/JS logic to replicate the WebView2 API on WASM to be inline with the efforts [here](https://github.com/unoplatform/uno/issues/4758)? Or is the UNO team planning on something different?

Thanks in advance for shedding some light

**Comment by jeromelaban** [comment]
In general, WebView use in a WebAssembly app is generally for JavaScript controls use, and you can use [the documentation provided here](https://platform.uno/docs/articles/interop/wasm-javascript-1.html). If you need to show actual web sites, you can create a wrapper around an `iframe` using the same techniques.

`iframe` instances have very significant security restrictions, making a potential `WebView` implementation less useful when compared to implementing specialized interop as suggested above.

**Reply by DierkDroth** [reply]
Thanks I'll look into ...

**Reply by DierkDroth** [reply]
@jeromelaban FYI the iFrame approach actually works pretty good...

However, I'm running into a problem: I'm using WebView2 as HTML editor which requires setting
`document.body.contenteditable="true"`

I understand that the iFrame equivalent would be
`<iFrame>.contentDocument.body.contenteditable="true"`
.. in order to have the <iFrame>.srcdoc editable.

However, it does not work in the UNO WASM context.

- I confirmed by the browser development tools, that the attribute is set correctly
- I verified by a simple HTML sample holding an iFrame that it should work in principle: I really can edit HTML in an iFrame
- I also tried setting .designMode = 'on': still same problem
- I also confirmed by the browser development tools that I could set the hosting .body to 'contenteditable="true"' and in fact the whole UNO WASM window became editable - just *not* the embedded iFrame

Now I have the suspicion that the UNO WASM glue code somehow prevents me having an editable HTML in an iFrame.
This is Windows10 + Chrome.

Could you please shed some light before I started ripping apart the UNO WASM glue code trying to find the culprit?

Thanks in advance

**Comment by DierkDroth** [comment]
Alright, I took the plunge and tried to understand what prevents "contenteditable='true'" working in iframes in the HTML code generated by UNO. 

Here are my findings:
- pls see minimal iframe sample project attached
- the HTML code generated by UNO is what I uploaded as IFrameEditable1.zip
- after some tinkering with HTML code I came up with IFrameEditable4.zip which does what I would expect: the text in the iframe is editable

There are 2 differences between both versions of the HTML:
- I replaced 'pointer-events: none' by 'pointer-events: auto' for the iframe
- I had to comment out a DIV **which comes after** (??) the iframe in question

So the real question is: what could I do make UNO generate the correct code so I had editable iframes?
Thanks in advance for looking into

[IframeEditable.zip](https://github.com/unoplatform/uno/files/7918353/IframeEditable.zip)
[IframeEditable1.zip](https://github.com/unoplatform/uno/files/7918356/IframeEditable1.zip)
[IframeEditable4.zip](https://github.com/unoplatform/uno/files/7918357/IframeEditable4.zip)

**Reply by dr1rrb** [reply]
@DierkDroth the pointer issue (`pointer-events: auto`) is going to be fixed by https://github.com/unoplatform/uno/pull/7929. 

For the divs at the bottom of the file: they are still present (like in all uno app), but I was able to select and edit text in your sample:
![image](https://user-images.githubusercontent.com/8635919/151625497-c49fe8ba-e19b-4a1e-90f7-bdf7625ecb49.png)

**Reply by DierkDroth** [reply]
@dr1rrb I'm sorry, I just ran another test:
- I replaced 'pointer-events: none' by 'pointer-events: auto' in Iframe1Editable.html **only** for the iframe element (= 1 occurence)
- left anything else untouched

It still would not work -> the iframe content still would not be editable. 
This is latest Chrome on latest Windows10.

Could you please describe what parts of the HTML text your fix would exactly change? Thanks

**Reply by jeromelaban** [reply]
As @dr1rrb mentioned, this is going to be fixed in uno 4.1.

**Reply by DierkDroth** [reply]
Thanks @jeromelaban. Looking forward to 4.1...

**Comment by DierkDroth** [comment]
@jeromelaban @all FYI

I was able to use the iframe approach consistently in my app whereever I was using WebView2 in the WindowsAppSdk version of the app.  I'm using WebView2 mostly for rendering and editing HTML content.

There is one limitation though: I had to apply custom JS code for the OAuth2/OpenId logic where I had to open a dedicated browser tab (+ window.postmessage communicaiton between browser tabs)
